consensus/istanbul/core/vrank.go

```diff
@@ -0,0 +1,218 @@
+	"github.com/klaytn/klaytn/consensus/istanbul"
+	"github.com/rcrowley/go-metrics"
+)
+
+type Vrank struct {
+	startTime            time.Time
+	view                 istanbul.View
+	committee            istanbul.Validators
+	threshold            time.Duration
+	commitArrivalTimeMap map[common.Address]time.Duration
+}
+
+var (
+	// VRank metrics
+	vrankFirstCommitArrivalTimeGauge           = metrics.NewRegisteredGauge("vrank/first_commit", nil)
+	vrankQuorumCommitArrivalTimeGauge          = metrics.NewRegisteredGauge("vrank/quorum_commit", nil)
+	vrankAvgCommitArrivalTimeWithinQuorumGauge = metrics.NewRegisteredGauge("vrank/avg_commit_within_quorum", nil)
+	vrankLastCommitArrivalTimeGauge            = metrics.NewRegisteredGauge("vrank/last_commit", nil)
+
+	vrankDefaultThreshold = "300ms"
+
+	vrank *Vrank
+)
+
+const (
+	vrankArrivedEarly = iota
+	vrankArrivedLate
+	vrankNotArrived
+)
+
+const (
+	vrankNotArrivedPlaceholder = -1
+)
+
+func NewVrank(view istanbul.View, committee istanbul.Validators) *Vrank {
+	threshold, _ := time.ParseDuration(vrankDefaultThreshold)
+	return &Vrank{
+		startTime:            time.Now(),
+		view:                 view,
+		committee:            committee,
+		threshold:            threshold,
+		commitArrivalTimeMap: make(map[common.Address]time.Duration),
+	}
+}
+
+func (v *Vrank) TimeSinceStart() time.Duration {
+	return time.Now().Sub(v.startTime)
+}
+
+func (v *Vrank) AddCommit(msg *istanbul.Subject, src istanbul.Validator) time.Duration {
+	if v.isTargetCommit(msg, src) {
+		t := v.TimeSinceStart()
+		v.commitArrivalTimeMap[src.Address()] = t
+		return t
+	}
+	return -1
+}
+
+func (v *Vrank) HandleCommitted(blockNum *big.Int) {
+	if v.view.Sequence.Cmp(blockNum) != 0 {
+		return
+	}
+
+	committedTime := v.TimeSinceStart()
+	if v.threshold > committedTime {
+		v.threshold = committedTime
+	}
+
+	vrankQuorumCommitArrivalTimeGauge.Update(int64(committedTime))
+	sum := int64(0)
+	for _, v := range v.commitArrivalTimeMap {
+		sum += int64(v)
+	}
+	avg := sum / int64(len(v.commitArrivalTimeMap))
+	vrankAvgCommitArrivalTimeWithinQuorumGauge.Update(avg)
+}
+
+// Log logs accumulated data in a compressed form
+func (v *Vrank) Log() (string, []string) {
+	var (
+		serialized = serialize(v.committee, v.commitArrivalTimeMap)
+		assessed   = assessBatch(serialized, v.threshold)
+		compressed = compress(assessed)
+		bitmap     = hex.EncodeToString(compressed)
+
+		lateCommits = make([]time.Duration, 0)
+		lastCommit  = time.Duration(0)
+	)
+
+	for _, t := range serialized {
+		if assess(t, v.threshold) == vrankArrivedLate {
+			lateCommits = append(lateCommits, t)
+		}
+	}
+
+	// lastCommit = max(lateCommits)
+	for _, t := range lateCommits {
+		if lastCommit < t {
+			lastCommit = t
+		}
+	}
+	if lastCommit != time.Duration(0) {
+		vrankLastCommitArrivalTimeGauge.Update(int64(lastCommit))
+	}
+
+	lateCommitsStrArr := encodeDurationBatch(lateCommits)
+
+	logger.Info("VRank", "bitmap", bitmap, "late", lateCommitsStrArr)
```

Comment:
To make parsing simpler, it may also print the View.

Reply:
good idea. added 46116f32